engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/KafkaAvroSink.scala

```diff
@@ -3,23 +3,95 @@ package pl.touk.nussknacker.engine.avro.sink
 import org.apache.flink.streaming.api.datastream.DataStreamSink
 import org.apache.flink.streaming.api.functions.sink.SinkFunction
 import pl.touk.nussknacker.engine.api.{InterpretationResult, LazyParameter, ValueWithContext}
 import pl.touk.nussknacker.engine.avro.encode.{BestEffortAvroEncoder, ValidationMode}
 import pl.touk.nussknacker.engine.avro.schemaregistry.{ExistingSchemaVersion, SchemaVersionOption}
 import pl.touk.nussknacker.engine.avro.serialization.KafkaAvroSerializationSchemaFactory
+import pl.touk.nussknacker.engine.avro.sink.KafkaAvroSink.KeyedObjectMapper
 import pl.touk.nussknacker.engine.flink.api.exception.{FlinkEspExceptionHandler, WithFlinkEspExceptionHandler}
 import pl.touk.nussknacker.engine.flink.api.process.{FlinkCustomNodeContext, FlinkSink}
 import pl.touk.nussknacker.engine.flink.util.keyed.{KeyedValue, KeyedValueMapper}
 import pl.touk.nussknacker.engine.kafka.{KafkaConfig, PartitionByKeyFlinkKafkaProducer, PreparedKafkaTopic}
 
-class KafkaAvroSink(preparedTopic: PreparedKafkaTopic, versionOption: SchemaVersionOption, key: LazyParameter[AnyRef], value: LazyParameter[AnyRef],
-                    kafkaConfig: KafkaConfig, serializationSchemaFactory: KafkaAvroSerializationSchemaFactory,
-                    schema: NkSerializableAvroSchema, runtimeSchema: Option[NkSerializableAvroSchema],
-                    clientId: String, validationMode: ValidationMode)
+object KafkaAvroSink {
+
+  import com.typesafe.scalalogging.LazyLogging
+  import org.apache.flink.api.common.functions.RichFlatMapFunction
+  import org.apache.flink.util.Collector
+  import pl.touk.nussknacker.engine.api.typed.typing
+  import pl.touk.nussknacker.engine.api.{Context, LazyParameter, LazyParameterInterpreter, ValueWithContext}
+  import pl.touk.nussknacker.engine.flink.api.process.{FlinkLazyParameterFunctionHelper, LazyParameterInterpreterFunction}
+  import pl.touk.nussknacker.engine.flink.util.keyed
+  import pl.touk.nussknacker.engine.flink.util.keyed.KeyedValue
+
+  import scala.util.control.NonFatal
+
+
+  // TODO FIXME: sanitize key!
+  class KeyedObjectMapper(protected val lazyParameterHelper: FlinkLazyParameterFunctionHelper,
+                          key: LazyParameter[AnyRef],
+                          fields: List[(String, LazyParameter[AnyRef])])
+    extends RichFlatMapFunction[Context, ValueWithContext[KeyedValue[AnyRef, AnyRef]]]
+      with LazyParameterInterpreterFunction with LazyLogging {
+    implicit def lazyParameterInterpreterImpl: LazyParameterInterpreter = lazyParameterInterpreter
+
+    lazy val buildObject: LazyParameter[Map[String, AnyRef]] = {
+      val emptyObj = Map.empty[String, AnyRef]
+      lazyObjectFieldsSequence
+        .map { list =>
+          list.foldLeft(emptyObj) { case (obj, field) =>
+            obj + (field._1 -> field._2)
+          }
+        }
+    }
+
+    override def flatMap(value: Context, out: Collector[ValueWithContext[KeyedValue[AnyRef, AnyRef]]]): Unit = {
```

Comment:
Let's use proper error handling... Like in EncodeAvroRecordFunction

Reply:
Done